File: app/src/main/java/de/kuschku/libquassel/primitives/types/QVariant.java
/*
 * QuasselDroid - Quassel client for Android
 * Copyright (C) 2016 Janne Koschinski
 * Copyright (C) 2016 Ken Børge Viktil
 * Copyright (C) 2016 Magnus Fjell
 * Copyright (C) 2016 Martin Sandsmark <martin.sandsmark@kde.org>
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.primitives.types;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.ArrayList;
import java.util.Set;

import de.kuschku.libquassel.primitives.QMetaType;
import de.kuschku.libquassel.primitives.QMetaTypeRegistry;
import de.kuschku.libquassel.syncables.SyncableRegistry;
import de.kuschku.libquassel.syncables.types.interfaces.QSyncableObject;

public class QVariant<T> {
    @Nullable
    public final T data;
    @Nullable
    public final QMetaType<T> type;

    @SuppressWarnings("unchecked")
    public QVariant(@NonNull T data) {
        if (data instanceof Set) {
            this.data = (T) new ArrayList((Set) data);
        } else {
            if (QMetaTypeRegistry.getTypeByObject(data) != null) {
                this.data = data;
            } else if (data instanceof QSyncableObject) {
                this.data = (T) SyncableRegistry.toVariantMap((QSyncableObject) data);
            } else {
                throw new IllegalArgumentException();
            }
        }

        this.type = QMetaTypeRegistry.getTypeByObject(this.data);
    }

    public QVariant(@NonNull QMetaType<T> type, @Nullable T data) {
        this.type = type;
        this.data = data;
    }

    public QVariant(@NonNull QMetaType.Type type, @Nullable T data) {
        this.type = QMetaTypeRegistry.getType(type);
        this.data = data;
    }

    public QVariant(@NonNull String typeName, @Nullable T data) {
        this.type = QMetaTypeRegistry.getType(typeName);
        this.data = data;
    }

    @Nullable
    public static <T> T orNull(@Nullable QVariant<T> data) {
        if (data == null) return null;
        else return data.data;
    }

    @NonNull
    public String toString() {
        return "QVariant(data=" + String.valueOf(this.data) + ", type=" + this.type.name + ")";
    }

    @NonNull
    public T or(@NonNull T ifNull) {
        return data == null ? ifNull : data;
    }

    @Nullable
    public T get() {
        return data;
    }
}

File: lib/src/main/java/de/kuschku/libquassel/protocol/QVariant.kt
/*
 * Quasseldroid - Quassel client for Android
 *
 * Copyright (c) 2019 Janne Mareike Koschinski
 * Copyright (c) 2019 The Quassel Project
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 3 as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package de.kuschku.libquassel.protocol

import de.kuschku.libquassel.protocol.primitive.serializer.Serializer

sealed class QVariant<T> constructor(val data: T?, val type: Type, val serializer: Serializer<T>) {
  class Typed<T> internal constructor(data: T?, type: Type, serializer: Serializer<T>) :
    QVariant<T>(data, type, serializer) {
    override fun toString() = "QVariant.Typed(${type.serializableName}, $data)"
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is Typed<*>) return false

      if (data != other.data) return false
      if (type != other.type) return false

      return true
    }

    override fun hashCode(): Int {
      var result = data?.hashCode() ?: 0
      result = 31 * result + type.hashCode()
      return result
    }
  }

  class Custom<T> internal constructor(data: T?, val qtype: QType, serializer: Serializer<T>) :
    QVariant<T>(data, qtype.type, serializer) {
    override fun toString() = "QVariant.Custom($qtype, $data)"
    override fun equals(other: Any?): Boolean {
      if (this === other) return true
      if (other !is Custom<*>) return false

      if (data != other.data) return false
      if (qtype != other.qtype) return false

      return true
    }

    override fun hashCode(): Int {
      var result = data?.hashCode() ?: 0
      result = 31 * result + qtype.hashCode()
      return result
    }
  }

  fun or(defValue: T): T {
    return data ?: defValue
  }

  companion object {
    @Suppress("UNCHECKED_CAST")
    fun <T> of(data: T?, type: Type): QVariant<T> {
      return QVariant.Typed(data, type, type.serializer as Serializer<T>)
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> of(data: T?, type: QType) =
      QVariant.Custom(data, type, type.serializer as Serializer<T>)
  }
}

inline fun <reified U> QVariant_?.value(): U? = this?.value<U?>(null)

inline fun <reified U> QVariant_?.value(defValue: U): U = this?.data as? U ?: defValue

inline fun <reified U> QVariant_?.valueOr(f: () -> U): U = this?.data as? U ?: f()

inline fun <reified U> QVariant_?.valueOrThrow(e: Throwable = NullPointerException()): U =
  this?.data as? U ?: throw e

inline fun <reified U> QVariant_?.valueOrThrow(e: () -> Throwable): U =
  this?.data as? U ?: throw e()
